service/src/main/java/bio/terra/tanagra/app/controller/CohortsV2ApiController.java

```diff
@@ -0,0 +1,204 @@
+    underlaysService.getUnderlay(body.getUnderlayName());
+    studyService.getStudy(studyId);
+
+    // Generate random 10-character alphanumeric strings for the new cohort and user-facing IDs.
+    String newCohortId = RandomStringUtils.randomAlphanumeric(10);
+    String newUserFacingCohortId = RandomStringUtils.randomAlphanumeric(10);
+
+    Cohort cohortToCreate =
+        Cohort.builder()
+            .studyId(studyId)
+            .cohortId(newCohortId)
+            .underlayName(body.getUnderlayName())
+            .userFacingCohortId(newUserFacingCohortId)
+            .version(Cohort.STARTING_VERSION)
+            .displayName(body.getDisplayName())
+            .description(body.getDescription())
+            .build();
+    cohortService.createCohort(cohortToCreate);
+    return ResponseEntity.ok(toApiObject(cohortService.getCohort(studyId, newUserFacingCohortId)));
+  }
+
+  @Override
+  public ResponseEntity<Void> deleteCohort(String studyId, String cohortId) {
+    accessControlService.throwIfUnauthorized(null, DELETE, COHORT, new ResourceId(cohortId));
+    cohortService.deleteCohort(studyId, cohortId);
+    return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortV2> getCohort(String studyId, String cohortId) {
+    accessControlService.throwIfUnauthorized(null, READ, COHORT, new ResourceId(cohortId));
+    return ResponseEntity.ok(toApiObject(cohortService.getCohort(studyId, cohortId)));
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortListV2> listCohorts(
+      String studyId, Integer offset, Integer limit) {
+    ResourceIdCollection authorizedCohortIds =
+        accessControlService.listResourceIds(COHORT, offset, limit);
+    List<Cohort> authorizedCohorts;
+    if (authorizedCohortIds.isAllResourceIds()) {
+      authorizedCohorts = cohortService.getAllCohorts(studyId, offset, limit);
+    } else {
+      authorizedCohorts =
+          cohortService.getCohorts(
+              studyId,
+              authorizedCohortIds.getResourceIds().stream()
+                  .map(ResourceId::getId)
+                  .collect(Collectors.toList()),
+              offset,
+              limit);
+    }
+
+    ApiCohortListV2 apiCohorts = new ApiCohortListV2();
+    authorizedCohorts.stream().forEach(cohort -> apiCohorts.add(toApiObject(cohort)));
+    return ResponseEntity.ok(apiCohorts);
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortV2> updateCohort(
+      String studyId, String cohortId, ApiCohortUpdateInfoV2 body) {
+    accessControlService.throwIfUnauthorized(null, UPDATE, COHORT, new ResourceId(cohortId));
+
+    List<CriteriaGroup> criteriaGroups =
+        body.getCriteriaGroups() == null
+            ? null
+            : body.getCriteriaGroups().stream()
+                .map(
+                    apiCriteriaGroup -> {
+                      // Generate random 10-character alphanumeric strings for each criteria group
```

Comment:
It would be much simpler to store entire cohorts as JSON or protocol buffers. Storing them in separate tables like this is more useful when actually querying against them (e.g. Tanagra indexes).

Reply:
That would be simpler and right now the backend could treat the criteria groups/criteria like a blob, since it's just passing it to the UI. But when we move the filter-building logic to the backend, then it does need to understand these objects, so I've stored it in a deserialized way with this in mind. That's generally the rule I've tried to follow -- if it's opaque to the backend, store as a blob, if the backend needs to understand it, then store it with columns/tables.

Reply:
I don't see the benefit of adding a de/serialization step to DB interactions. I also think it's not a good idea to couple serialization formats throughout the stack (i.e. UI relies on same serialization structures as the internal objects as the backend DB) -- it causes a lot of headache if we ever need to update one of those or deal with upgrades.

Reply:
Per offline conversation, going to leave as is for now. In the future, we may consider storing JSON, but need a good way to handle schema upgrades first (maybe OpenAPI generated classes? protobufs?). Would be good to think about this after upgrading the Java version (so we can consider Java records as an option) and reevaluating the underlay config schema definition.